Instructions for additional track/line following lab exercises (#22)
lf-lang/embedded-lab: src/Line.md

# 9 Line Following

> The goal of this exercise is to do LF programming using the Pololu robot to execute a specified task, namely line follwoing, where the robot detects the line and follows the line until it reaches the end of line.

Reviewer: ```suggestion
The goal of this exercise is to program the Pololu robot in LF to perform line following, where the robot detects a line and follows it until it reaches the end.
```

For this lab, you will need a suitable line like the one here (made using black ducktape and a white foam board):

<img src="img/Line.jpeg">

## 9.1 Prelab

<!---
NOTE: The prelabs may change according to the ordering of the lab.
-->

- Review [Section 6.5, Line and bump sensors](https://www.pololu.com/docs/0J86/6.5) of the [Pololu 3pi+ 2040 robot User's Guide](https://www.pololu.com/docs/0J86).

> - Review the section for the motors from the previous lab. This lab will test you how to adjust power to maneuver curves and sharp turns.

Reviewer: ```suggestion
- Review the section for the motors from the previous lab. This lab will test how you adjust power to maneuver curves and sharp turns.
```

> - Review line sensing, the same calibration should be done at the beginning of this excersie as well. However, there is a small difference in how the sensor is calibrated.

Reviewer: Where is "line sensing"?  Hyperlink here?  Also:
```suggestion
- Review line sensing. The same calibration should be done at the beginning of this exercise, however, with a small difference, explained below.
```

Author: Thank you for catching this! Fixed this in 88158b927f042547166f98051f2196eb9ab172d0.

>    For this lab, you are required to indicate where the robot is going to move as the sensor moves out of the line in order for the robot to realign itself with the track.

Reviewer: The above sentence is not clear.  What does it mean to "indicate where the robot is going"?

Reviewer 2: I have changed it to "indicate where the robot is planning or programmed to move" in the latest commit. I have also added a brief explanation in brackets.

  
**Questions**

1. Can we use the gyroscope to make the robot navigate the curves?
2. Can we use the line sensors on the edges, i.e., sensor 1 and 5 to detect angles and curves? 

## 9.2 Line Following

**Checkoff:** Calibrate your line sensors to detect the "out-of-track" area.

> Write LF code to make your robot follow the designated path. The trick is to make sure the robot's motor is calibrated perfectly to make the turns, adjust the speed of the robot.

Reviewer: This is worrisome. Perfect calibration is not possible.

Reviewer 2: Oh yes you're right, that was worded incorrectly, I have made the change as :  "make sure the speed of the robot's motor enables it to maneuver the turns, adjust the speed of the robot if necessary"

> **HINT:** 0.1f would be perfect to start, you may increase or decrease the speed eventually. The goal is functionality.

Reviewer: I think "perfect" can't be right here.  You mean that this is a reasonable guess?  Also, there are two motors.  What does it mean to calibrate them?  Are you referring to speed here rather than calibration?

Reviewer 2: Yes I was referring to the speed, I have changed it to: "0.1f would be a reasonable starting point for the speed".
Thank you for the comments, I have made the changes necessary.

> You may build a track at your own home, to test your code. The important thing to keep in mind is that the track must be smooth and the robot should not get stuck somewhere in between. The key thing is that the track and the surroundings must have a significant contrast in terms of the color. You are free to test your code on that track before you submit the code for it. The robot should align itself without any external help, even if it starts out skewered with respect to the track. The robot should stop at the end of the track or keep continuing if the track is a loop. You are free to include any module of your choice to make sure the robot is precise in keeping itself on track, the robot should have enough intelligence or functionality to keep itself on track. If it reaches the end of the track the robot should stop. The robot will be tested on a different track from the one you developed your code with, to ensure uniformity. Please put your solution in a file called `LineFollowSolution.lf`.

Reviewer: ```suggestion
You may build a track at your own to test your code. The important thing to keep in mind is that the track must be smooth and the robot should not get stuck somewhere on the track. The track and the surroundings must have a significant contrast in color. You are free to test your code on that track before you submit the code. The robot should align itself without any external help even if it starts out skewed with respect to the track. The robot should stop at the end of the track or continue indefinitely if the track forms a loop. The robot should have enough intelligence or functionality to keep itself on track. The robot will be tested on a different track from the one you developed your code with, to ensure adaptability. Please put your solution in a file called `LineFollowSolution.lf`.
```

> **Checkoff:** Show your robot driving to the end of the line, handling the curves smoothly.

Reviewer: ```suggestion
**Checkoff:** Show your robot driving to the end of the line and handling the curves smoothly.
```

- Review [Line Sensing](./Hill.md#82-line-sensing). The same calibration should be done at the beginning of this exercise, however, with a small difference, explained below.
 
  For this lab, you are required to indicate where the robot is planning or programmed to move when the sensor detects an out-of-track area in order for the robot to realign itself with the track (For instance, if the right-hand-side sensors of the robot detect an out-of-track area the print on display should read "left" and vice-versa).  

**Questions**

1. Can we use the gyroscope to make the robot navigate the curves?
2. Can we use the line sensors on the edges, i.e., sensors 1 and 5, to detect angles and curves? 

## 9.2 Track Following

**Checkoff:** Calibrate your line sensors to detect the "out-of-track" area.

Write LF code to make your robot follow the designated path. The trick is to make sure the speed of the robot's motor enables it to maneuver the turns and adjust the speed of the robot if necessary.

**HINT:** 0.1f would be a reasonable starting point for the speed. You may increase or decrease the speed eventually. The goal is functionality.

> You may build a track at your own home to test your code. The important thing to keep in mind is that the track must be smooth, and the robot should not get stuck somewhere in between. The key thing is that the track and the surroundings must have a significant contrast in terms of color. You are free to test your code on that track before you submit the code for it. The robot should align itself without any external help, even if it starts out skewered with respect to the track. The robot should stop at the end of the track or keep continuing if the track is a loop. You are free to include any module of your choice to make sure the robot is precise in keeping itself on track, and the robot should have enough intelligence or functionality to keep itself on track. If it reaches the end of the track, the robot should stop. The robot will be tested on a different track from the one you developed your code with to ensure uniformity. Please put your solution in a file called `TrackFollowSolution.lf`.

Reviewer: ```suggestion
You may build a track at your own home to test your code. The important thing to keep in mind is that the track must be smooth, and the robot should not get stuck somewhere in between. The key thing is that the track and the surroundings must have a significant contrast in terms of color. You are free to test your code on that track before you submit the code for it. The robot should align itself without any external help, even if it starts out skewed with respect to the track. The robot should stop at the end of the track or keep continuing if the track is a loop. You are free to include any module of your choice to make sure the robot is precise in keeping itself on track, and the robot should have enough intelligence or functionality to keep itself on track. The robot will be tested on a different track from the one you developed your code with to ensure uniformity. Please put your solution in a file called `TrackFollowSolution.lf`.
```
